libqtile/core/manager.py

```diff
@@ -441,13 +465,13 @@ def add_group(self, name, layout=None, layouts=None, label=None):
             return True
         return False
 
-    def delete_group(self, name):
+    def delete_group(self, name: str) -> None:
         # one group per screen is needed
         if len(self.groups) == len(self.screens):
             raise ValueError("Can't delete all groups.")
         if name in self.groups_map.keys():
             group = self.groups_map[name]
-            if group.screen and group.screen.previous_group:
+            if group.screen and hasattr(group.screen, "previous_group"):
```

Comment:
This looks like a bit odd, is there a reason the old way of having `previous_group` be an optional would not work with the mypy tests?

Reply:
It does look a bit odd, but it's either this or peppering `assert`s throughout the code, which IMO makes things unclear as to _why_ asserts are lying around the code, as well as messy and executing unnecessary code. Same with `self.qtile: Qtile = None` that we have on a number of classes, which results in us having lots of lines like here: https://github.com/qtile/qtile/blob/master/libqtile/backend/x11/core.py#L274

The `hasattr` lines are only required in the few places where these attributes can be accessed before getting populated, which in the case of `self.qtile` on things like windows pretty much never happens, but it can happen on `Screen` as you see here with `self.previous_group`.

To be honest I'm not sure if this is just a weakness in Python, or mypy, and it's probably just down to preference, so I'm happy to do whatever with these ones.